src/net.rs

```diff
@@ -25,6 +25,48 @@ use tls::wrap_tls;
 /// This constant should be more than the largest expected RTT.
 pub(crate) const TIMEOUT: Duration = Duration::from_secs(60);
 
+pub(crate) async fn update_connection_history(
+    context: &Context,
+    alpn: &str,
+    host: &str,
+    port: u16,
+    addr: &str,
+    now: i64,
+) -> Result<()> {
+    context
+        .sql
+        .execute(
+            "INSERT INTO connection_history (host, port, alpn, addr, timestamp)
+             VALUES (?, ?, ?, ?, ?)
+             ON CONFLICT (host, port, alpn, addr)
```

Comment:
So anyway this table may grow without limit, particularly if the address is dynamic, and the `dns_cache` table as well, though this is not the case for most hosts. Also it makes sense to clean up these tables after a reconfiguration, though also this is not the case for most users. `dns_cache` entries older than 1 month are not used anyway, maybe just clean up everything older than 1 month?